modules/element-manager/ajax.php

```diff
@@ -68,6 +79,48 @@ public function ajax_get_admin_page_data() {
 		wp_send_json_success( $data );
 	}
 
+	private function get_element_manager_permissions_promotion(): array {
+		$promotion_data = [
+			'pro' => [
+				'text' => esc_html__( 'Upgrade Now', 'elementor' ),
+				'url' => self::FREE_TO_PRO_PERMISSIONS_PROMOTION_URL,
+			],
+			'advanced' => [
+				'text' => esc_html__( 'Upgrade Now', 'elementor' ),
+				'url' => self::PRO_TO_ADVANCED_PERMISSIONS_PROMOTION_URL,
+			],
+		];
+
+		$filtered_data = apply_filters( 'elementor/element_manager/admin_app_data/promotion_data/manager_permissions', $promotion_data );
+
+		foreach ( $promotion_data as $key => $data ) {
+			$filtered_data[ $key ] = $this->validate_promotion_data( $filtered_data[ $key ], $data );
+		}
+
+		return $filtered_data;
+	}
+
+	private function get_element_manager_promotion(): array {
+		$promotion_data = [
+			'text' => esc_html__( 'Upgrade Now', 'elementor' ),
+			'url' => self::ELEMENT_MANAGER_PROMOTION_URL,
+		];
+
+		$filtered_data = apply_filters( 'elementor/element_manager/admin_app_data/promotion_data/element_manager', $promotion_data );
+
+		return $this->validate_promotion_data( $filtered_data, $promotion_data );
+	}
+
+	private function validate_promotion_data( $filtered_data, $promotion_data ) {
```

Comment:
validate_promotion_data replaces the values?

Reply:
maybe check that the returned array is not too big ?

Reply:
> maybe check that the returned array is not too big ?

https://github.com/elementor/elementor/pull/25088/files#diff-f2b395a7fa05fe6f81da6c9c5b1616122fca1495d0a19bb07b11146f32dce463R239

Reply:
@TzviRabinovitch @louiswol94 
I don't think checking the array length is a good solution.
What if the array length is the same but the keys are different?
I see two possible solutions:
1. Maybe we should obligate the end user to use all the keys in the base array.
2. We can map the base array and change the values only on the existing keys.

WDYT?

Reply:
if the keys are different than we end up with x2 the size which is "ok" , the concern is that it might be x1000